HIPs/firefly.md

```diff
@@ -0,0 +1,374 @@
+  - Pin evidence of that transfer to a shared ledger, with identity + locked sequence
+  - Perform off-chain only transfers for latency intensive operations (state channels)
+  - Record off-chain delivery receipts
+- Document handling
+  - Business processes involve the exchange and attestation of many documents
+  - These need to flow between parties, and often must remain private
+  - Parties have access to different pieces of data, within a single transaction
+
+#### Network registry
+
+Permissioning is a core tenant of enterprise multi-party systems, as is the ability to scale and grow the network dynamically over time.
+
+The FireFly network registry provides a pluggable governance framework for onboarding of new organizations into the network. Then a distribution framework to publish and distribute the node/endpoint connectivity information (on-chain and off-chain) for that organization, as well to allow that organization to publish identities to the network that are used for transactions.
+
+The registry publication system uses the same model for distribution and sequencing of new data, as applications that build on top of FireFly. This means the same state of the registry can be deterministically reached by all members, while each member has their own fast local cache of data.
+
+#### Integration of core blockchain constructs
+
+Core blockchain constructs are composable building blocks of multi-party solutions, and can coexist seamlessly with traditional private messaging and back-office system integration.
+
+- Tokens
+  - Non-fungible tokens (NFTs) - universally unique
+  - Fungible tokenized value exchange - total conversation of value
+- Signatures
+  - Attestation to data, at a particular point in time
+- Multi-party event sequencing
+  - The game-changing feature of blockchain for enterprise multi-party systems
+- Deterministic logic
+  - Custom on-chain smart contracts
+  - Trusted execution environments (TEEs)_ - plug point_
+  - Zero-knowledge Proofs (ZKPs) - _plug point_
+  - Multi-party Compute (MPC) - _plug point_
+
+#### Coordination of on-chain and off-chain interactions
+
+Sometimes underestimated is the importance, and complexity, of synchronizing the exchange of on-chain and off-chain data. Yet tackling this problem is almost universal in enterprise multi-party system development.
+
+FireFly provides a rugged, enterprise-grade runtime and simple event-driven REST API model, so that developers can focus on solving interesting business problems - rather than re-building plumbing. Second order problems are also addressed, like high throughput through batching of on-chain transactions, and providing n-way application scale while retaining deterministic ordering.
+
+The pluggable architecture allows the underlying infrastructure components to be swapped out, without changing the application. That includes the database, blob store, private messaging technology, and even the blockchain itself.
```

Comment:
which are the minimum requirements for a blockchain to be plugged in FireFly? Would Bitcoin work?

Reply:
But which are the requirements? Can Bitcoin be plugged?

Reply:
ah got you, so, the abstraction of the blockchain layer from the FireFly design is the following:
1. global ordering/time stamping
2. atomic broadcast
3. double spend prevention/global uniqueness
4. event publishing

Each is an optional feature available via the FireFly core API, then implemented via the plugin interface. Using an example, when Alice and Bob are exchanging a bundle of documents (say for global trade), they can:
- send the documents one time (say from Alice to Bob), both get a receipt with both parties signatures. end of story
- the two parties send modifications to the documents after the initial send, version history must be determined, they solve this by pinning their send operation with a blockchain tx (tied to the docs with the content hash). this utilizes 1) above. FireFly has an option parameter in the send API to pin to blockchain or not.
- Bob after obtaining the documents, wants to sell the entitlement in the trade marketplace, he saves a new document that describes high level terms of the documents in IPFS, and broadcast its reference to the whole network. Everybody can decide to look it up in their own IPFS node and follow up with a bid or ignore it. this utilizes 2) above. FireFly has an API that does the broadcast of metadata.
- a 3rd party, Charlie, is the eventual inspector of the authenticity of the documents (say Charlie is the ship captain that needs to decide whether to release the cargo after inspecting the documents) and the owner's entitlement. this requires the pinning transaction of the documents transfer to mint a non-fungible token that is bound to the documents (via content hash). only the individual who can demonstrate ownership of the private key corresponding to the owner's identity of the token can take delivery of the cargo. this utilizes 3) above. specific API designs for this part is to be finalized.

4) for event publishing is utilized in all scenarios touching the blockchain.

in summary, FireFly's philosophy is that the majority of multi-party system solutions rely on the blockchain layer for the above set of functionalities, which are common to practically all blockchain implementations. this is the reason for pluggability on the blockchain layer.

Reply:
Very interesting, this confirms that blockchain is used essentially as time-stamping service. 
Given that you token part has not been finalised (a big piece missing, I would say), it is difficult to say anything there. 
Not to mention the security/privacy implications of connecting the hashes to the tokens. 
It is also not clear to me, why Charlie should trust the piece of information that he sees. Probably, as in Corda, Charlie would need a train of transactions from the creation of the state to the its latest version passed to Charlie. Another potential source of leakage here. Is FireFly for Ethereum what the Flows are for the Notary? I have the impression that Firefly is trying to solve the same problems. Interesting.

What about the programming model? What are these transactions? Do you have a blockchain agnostic transaction specification?

You mentioned existing FireFly API, can you point to them? I have searched the repo but I found only the blockchain.Plugin interface. Nothing for tokens or a transaction definition. I must be missing something.

Thanks Jim :)

Reply:
Actually, it is interesting that you didn't mention immutability. A fundamental property of any blockchain.

Reply:
Jim, if I understand correctly, from looking to the code, the nodes do not assemble blockchain transactions. They exchange data and then using the `blockchain.Plugin` interface they timestamp hashes for traceability. Is this correct?
Which business logic do the parties execute? How?

Thanks again, Jim :)

Reply:
> Given that you token part has not been finalised (a big piece missing, I would say), it is difficult to say anything there.
Not to mention the security/privacy implications of connecting the hashes to the tokens.

Note that the 4 sample use cases are meant to be independent of each other. You should notice that tokens usage is not necessary in most of the sample use cases. It's a reflection of the enterprise multi-party system use case requirements in the real world. that said, we are working on a design and will be sharing with the community soon.

> why Charlie should trust the piece of information that he sees
Charlie would open the shipping company's iPad App, which is connected to the backend administered by his own company, which is connected to a blockchain node administered by his own company. The transfer history of the token recorded in the ledger will give Charlie confidence that Bob is indeed the current owner of the subject documents.

Token is a popular and proven construct for double spending protection but like you said it does have privacy implications. in some use cases network-wide ownership verifiability is desirable, such as a cooperative improvement of industry-wide data (e.g practitioner information).

> Is FireFly for Ethereum what the Flows are for the Notary?
FireFly is a blockchain-agnostic approach to solving the multi-party flow problem. FireFly is easier to program with than Corda's Flow framework. In a majority of the use cases there's no need to be concerned with onchain logic at all. it's also more versatile than the flow framework as it supports not only structured data but also files/blobs

Reply:
> the nodes do not assemble blockchain transactions. They exchange data and then using the blockchain.Plugin interface they timestamp hashes for traceability. Is this correct?

correct but with the clarification that blockchain serves a much more important role than traceability, which is global ordering that would otherwise be impossible in multi-party systems. as illustrated in the 2nd bullet above in the sample scenarios.

> Which business logic do the parties execute? How?

FireFly has no opinion about what business logic should be executed and how, but instead provide support for these different ways to process the messages/data that FireFly helps the members to exchange:
- member-specific local processing in the app layer, once events are pushed from FireFly to the client apps via event stream channels (eg. websocket)
- onchain smart contract, the FireFly core's data aggregator will be directed to submit the data (once it has been received from the offchain channel and matched with the hashes from the blockchain events) to the target contract to process. this is also a plug point for generating and sending zk proofs instead of the raw data to the contract to verify
- externally managed TEE cluster (as in Avalon style), the data aggregator will submit encrypted data to an enclave to process